DasharoPayloadPkg/Library/FmpDeviceSmmLib/FmpDeviceSmmLib.c

```diff
@@ -956,9 +958,13 @@ FmpDeviceSetImageWithStatus (
     return Status;
   }
 
-  ReadBuffer = AllocatePool (BlockSize);
-  if (ReadBuffer == NULL) {
-    DEBUG ((DEBUG_ERROR, "%a(): failed to allocate read buffer\n", __FUNCTION__));
+  CurrentImage = ReadCurrentFirmware ();

```

Comment:
> Made this read to move progress by about 5%

That part is resolved.

> Turned out that first 4 MiB of flash aren't mmapped, only "bios" region is. Not sure how to resolve this properly, locally I've added 4 * MiB to size at https://github.com/Dasharo/coreboot/blob/bd294ab1d22b4010fafa0120e2eda26964316096/src/soc/intel/common/block/fast_spi/mmap_boot.c#L123 which made the whole flash available to coreboot.

Interesting, especially since it then creates a device for whole `CONFIG_ROM_SIZE` in https://github.com/Dasharo/coreboot/blame/bd294ab1d22b4010fafa0120e2eda26964316096/src/soc/intel/common/block/fast_spi/mmap_boot.c#L134. There is no discussion as to size on gerrit, so perhaps this mismatch wasn't deliberate, or maybe it was done so that SPI addresses are correct.

I think the best option would be to add another Kconfig for mmaping whole ROM instead of just BIOS region, and make it selected when capsules are enabled.

Reply:
Anyway, this is coreboot issue, it doesn't block merging this PR.